shell/platform/android/io/flutter/view/FlutterMain.java

```diff
@@ -343,18 +378,44 @@ private static void initResources(Context applicationContext) {
         }
     }
 
+    private static Set<String> listLibs(Context applicationContext) {
+        ApplicationInfo applicationInfo;
+        try {
+            applicationInfo = applicationContext
+                .getPackageManager()
+                .getApplicationInfo(applicationContext.getPackageName(), PackageManager.GET_META_DATA);
+        } catch (PackageManager.NameNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+        ImmutableSetBuilder builder = ImmutableSetBuilder.<String>newInstance();
+        File[] files = new File(applicationInfo.nativeLibraryDir).listFiles();
+
+        for (File file : files) {
+            builder.add(file.getName());
+        }
+        return builder.build();
+    }
+
     private static void initAot(Context applicationContext) {
-        Set<String> assets = listAssets(applicationContext, "");
+        Set<String> assets;
+
+        if (sIsPrecompiledAsBlobLibraries) {
+            assets = listLibs(applicationContext);
```

Comment:
Is this call to `listLibs` redundant with the `listLibs` done in `initConfig`?

I'd suggest moving the `initConfig` library check code into `initAot`.  `initAot` can check for the presence of the blob libraries.  If they exist, then set `sIsPrecompiledAsBlobs = true` and rewrite the `sAotVm`/`sAotIsolate` filenames.

Instead of a separate `sIsPrecompiledAsBlobLibraries` flag, add a field that stores the directory path where the blobs live.  If the app is using blob libraries, this is the `nativeLibraryPath`.  If it's using blob assets, it will be the `PathUtils.getDataDirectory` path where the `ResourceExtractor` stores them.  Then pass this path as the `aot-snapshot-path` flag to the engine.

Reply:
I simplified this logic a bit. Matt's work is going to help a lot to make this code simpler and testable. :) PTAL